examples/simple_intersection/example_intersection.cpp

```diff
@@ -0,0 +1,168 @@
+  static KOKKOS_FUNCTION int size(Boxes<DeviceType> const &boxes)
+  {
+    return boxes.size();
+  }
+  static KOKKOS_FUNCTION auto get(Boxes<DeviceType> const &boxes, int i)
+  {
+    return boxes.get_box(i);
+  }
+};
+
+// For performing the queries given a Boxes object, we need to define memory
+// space, how to get the total number of queries, and what the query with index
+// i should look like. Since we are using self-intersection (which boxes
+// intersect with the given one), the functions here very much look like the
+// ones in ArborX::AccessTraits<Boxes<DeviceType>, ArborX::PrimitivesTag>.
+template <typename DeviceType>
+struct ArborX::AccessTraits<Boxes<DeviceType>, ArborX::PredicatesTag>
+{
+  using memory_space = typename DeviceType::memory_space;
+  static KOKKOS_FUNCTION int size(Boxes<DeviceType> const &boxes)
+  {
+    return boxes.size();
+  }
+  static KOKKOS_FUNCTION auto get(Boxes<DeviceType> const &boxes, int i)
+  {
+    return intersects(boxes.get_box(i));
+  }
+};
+
+// Now that we have encapsulated the objects and queries to be used within the
+// Boxes class, we can continue with performing the actual search.
+int main()
+{
+  Kokkos::initialize();
+  {
+    using ExecutionSpace = Kokkos::DefaultExecutionSpace;
+    using MemorySpace = typename ExecutionSpace::memory_space;
+    using DeviceType = Kokkos::Device<ExecutionSpace, MemorySpace>;
+    ExecutionSpace execution_space;
+
+    std::cout << "Create grid with bounding boxes" << '\n';
+    Boxes<DeviceType> boxes(execution_space);
+    std::cout << "Bounding boxes set up." << '\n';
+
+    std::cout << "Creating BVH tree." << '\n';
+    ArborX::BVH<MemorySpace> const tree(execution_space, boxes);
+    std::cout << "BVH tree set up." << '\n';
+
+    std::cout << "Starting the queries." << '\n';
+    // The query will resize indices and offsets accordingly
+    Kokkos::View<int *, MemorySpace> indices("indices", 0);
+    Kokkos::View<int *, MemorySpace> offsets("offsets", 0);
+
+    ArborX::query(tree, execution_space, boxes, indices, offsets);
+    std::cout << "Queries done." << '\n';
+
+    std::cout << "Starting checking results." << '\n';
+    auto offsets_host =
+        Kokkos::create_mirror_view_and_copy(Kokkos::HostSpace{}, offsets);
+    auto indices_host =
+        Kokkos::create_mirror_view_and_copy(Kokkos::HostSpace{}, indices);
+
+    unsigned int const n = boxes.size();
+    if (offsets_host.size() != n + 1)
+      Kokkos::abort("Wrong dimensions for the offsets View!\n");
+    for (int i = 0; i < static_cast<int>(n + 1); ++i)
+      if (offsets_host(i) != i)
+        Kokkos::abort("Wrong entry in the offsets View!\n");
+
+    if (indices_host.size() != n)
+      Kokkos::abort("Wrong dimensions for the indices View!\n");
+    for (int i = 0; i < static_cast<int>(n); ++i)
+      if (indices_host(i) != i)
+        Kokkos::abort("Wrong entry in the indices View!\n");
+    std::cout << "Checking results successful." << '\n';
+  }
```

Comment:
You could probably write it more compact with `std::vector`, `std::iota` and `operator==`, but no biggie.